ceno_zkvm/src/instructions/riscv/mul.rs

```diff
@@ -308,16 +350,32 @@ impl<E: ExtensionField, I: RIVInstruction> Instruction<E> for MulhInstructionBas
                 rd_signed.assign_instance(instance, lk_multiplicity, &rd_val)?;
 
                 let prod = (rs1_s as i64) * (rs2_s as i64);
-                ((prod as u64) % (1u64 << BIT_WIDTH)) as u32
+                // only take the low part of the product
+                prod as u32
             }
             MulhSignDependencies::UU { constrain_rd } => {
                 // assign nonzero value (u32::MAX - rd)
                 let rd_f = E::BaseField::from(rd as u64);
-                let avoid_f = E::BaseField::from((1u64 << BIT_WIDTH) - 1);
+                let avoid_f = E::BaseField::from(u32::MAX.into());
                 constrain_rd.assign_instance(instance, rd_f, avoid_f)?;
 
                 let prod = rs1_val.as_u64() * rs2_val.as_u64();
-                (prod % (1u64 << BIT_WIDTH)) as u32
+                // only take the low part of the product
+                prod as u32
+            }
+            MulhSignDependencies::LL { constrain_rd } => {
+                let prod = rs1_val.as_u64() * rs2_val.as_u64();
+                let prod_lo = prod as u32;
+                assert_eq!(prod_lo, rd);
+
+                let prod_hi = (prod >> BIT_WIDTH) as u32;
```

Comment:
Very, very minor cosmetic suggestion to decrease the number of casts the reader has to reason through:

```rust
let prod_hi = (prod >> BIT_WIDTH);
...
E::BaseField::from(prod_hi),
...
prod_hi as u32
```

Reply:
addressed in [11df823](https://github.com/scroll-tech/ceno/pull/660/commits/11df82393415723d06aab78410cf4d8665f76254)